contracts/AelinUpFrontDeal.sol

```diff
@@ -400,41 +410,47 @@ contract AelinUpFrontDeal is AelinERC20, MinimalProxyFactory, IAelinUpFrontDeal
         }
     }
 
+    function claimUnderlyingMutlipleEntries(uint256[] memory _indices) external {
+        for (uint256 i = 0; i < _indices.length; i++) {
+            _claimUnderlying(msg.sender, _indices[i]);
+        }
+    }
+
     /**
-     * @dev purchaser calls after the purchasing period to claim underlying deal tokens
+     * @dev ERC721 deal token holder calls after the purchasing period to claim underlying deal tokens
      * amount based on the vesting schedule
      */
-    function claimUnderlying() external lock purchasingOver passMinimumRaise {
-        uint256 underlyingDealTokensClaimed = claimableUnderlyingTokens(msg.sender);
-        require(underlyingDealTokensClaimed > 0, "no underlying ready to claim");
+    function claimUnderlying(uint256 _tokenId) external {
+        _claimUnderlying(msg.sender, _tokenId);
+    }
+
+    function _claimUnderlying(address _owner, uint256 _tokenId) internal purchasingOver passMinimumRaise {
+        require(ownerOf(_tokenId) == _owner, "must be owner to claim");
+        uint256 claimableAmount = claimableUnderlyingTokens(_tokenId);
+        require(claimableAmount > 0, "no underlying ready to claim");
         address _underlyingDealToken = dealData.underlyingDealToken;
-        amountVested[msg.sender] += underlyingDealTokensClaimed;
-        _burn(msg.sender, underlyingDealTokensClaimed);
-        totalUnderlyingClaimed += underlyingDealTokensClaimed;
-        IERC20(_underlyingDealToken).safeTransfer(msg.sender, underlyingDealTokensClaimed);
-        emit ClaimedUnderlyingDealToken(msg.sender, _underlyingDealToken, underlyingDealTokensClaimed);
+        tokenDetails[_tokenId].lastClaimedAt = block.timestamp;
+        amountVested[msg.sender] += claimableAmount;
+        totalUnderlyingClaimed += claimableAmount;
+        IERC20(_underlyingDealToken).safeTransfer(_owner, claimableAmount);
+        emit ClaimedUnderlyingDealToken(_owner, _underlyingDealToken, claimableAmount);
     }
 
     /**
-     * @dev a view showing the amount of the underlying deal token a purchaser can claim
-     * @param _purchaser address to check the quantity of claimable underlying tokens
+     * @dev a view showing the amount of the underlying deal token a ERC721 deal token holder can claim
+     * @param _tokenId the token ID to check the quantity of claimable underlying tokens
      */
-    function claimableUnderlyingTokens(address _purchaser) public view purchasingOver returns (uint256) {
-        uint256 _vestingPeriod = dealConfig.vestingPeriod;
-        uint256 precisionAdjustedUnderlyingClaimable;
 
+    function claimableUnderlyingTokens(uint256 _tokenId) public view returns (uint256) {
```

Comment:
it is just a view but it will error if you pass it a token that doesn't exist. also if the balance returned is 0 it feels like there is a way to not run so much of the logic

Reply:
right because schedule would be undefined but its assumed that the field has values

Reply:
yeah it's true maybe I can add something like 

` if (schedule.lastClaimedAt > 0) {}` at the top of the function